### ISSUE TITLE
Possible core generation race condition when multiple processes crash

### ISSUE THREAD

**Issue by springmeyer** [mapbox/logbt#5]
It is standard for the `core_pattern` to use `%p` to name core files by `PID`. Currently we set the `core_pattern` to `/tmp/logbt-coredump`. This means that if multiple processes crash we at best only get a core dump for the last one. At worst we'll get corrupted coredumps and broken backtraces.

Here is a demo of the problem:

```bash
# create program that loops forever
echo "int main() { while(true) {} }" > forever.cpp
g++ -o forever forever.cpp

export COREFILE=/tmp/logbt-coredump
echo $COREFILE > /proc/sys/kernel/core_pattern
ulimit -c unlimited

function stop_it() {
    kill -s SEGV `pidof $1` && gdb $(pwd)/$1 $COREFILE -ex "set pagination 0" -ex "thread apply all bt" --batch
}

# starting one loop and manually crashing should work
# as a single coredump and backtrace is generated
./forever & stop_it forever
```

That will result in a nice clean backtrace (that is associated with the correct program):

```
Core was generated by `./forever'.
Program terminated with signal 11, Segmentation fault.
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3282):
#0  0x00000000004004b8 in main ()
[10]   Segmentation fault      (core dumped) ./forever
```

But if multiple core dumps are generated they may clobber each other and backtrace generation may fail:

```
# make several copies
cp forever forever2
cp forever forever3
cp forever forever4
cp forever forever5

# start all the loops
./forever & ./forever2 & ./forever3 & ./forever4 & ./forever5 &

# manually crash and generate backtrace for each
stop_it forever & stop_it forever2 & stop_it forever3 & stop_it forever4 & stop_it forever5

```

Multiple loops clobber each other. In this case backtrace data is only displayed for `forever4`. And some backtraces cannot be generated as gdb exits with the error of `core file may not match specified executable file`. Full output was:

```
$ ./forever & ./forever2 & ./forever3 & ./forever4 & ./forever5 &
[10] 3316
[11] 3317
[12] 3318
[13] 3319
[14] 3320
ubuntu@ip-10-22-34-158:~$ stop_it forever & stop_it forever2 & stop_it forever3 & stop_it forever4 & stop_it forever5
[15] 3321
[16] 3322
[17] 3324
[18] 3326

warning: core file may not match specified executable file.
[New LWP 3319]

warning: Can't read pathname for load map: Input/output error.
[New LWP 3319]

warning: core file may not match specified executable file.
[New LWP 3319]

warning: Can't read pathname for load map: Input/output error.

warning: Can't read pathname for load map: Input/output error.
Core was generated by `./forever4'.
Program terminated with signal 11, Segmentation fault.
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3319):
#0  0x00000000004004b8 in main ()
Core was generated by `./forever4'.
Program terminated with signal 11, Segmentation fault.
Core was generated by `./forever4'.
Program terminated with signal 11, Segmentation fault.
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3319):
#0  0x00000000004004b8 in main ()
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3319):
#0  0x00000000004004b8 in main ()
[10]   Segmentation fault      (core dumped) ./forever
[11]   Segmentation fault      (core dumped) ./forever2
[12]   Segmentation fault      (core dumped) ./forever3
[13]   Segmentation fault      (core dumped) ./forever4
[14]   Segmentation fault      (core dumped) ./forever5
[17]   Done                    stop_it forever3
ubuntu@ip-10-22-34-158:~$ 
warning: core file may not match specified executable file.
[New LWP 3319]

warning: Can't read pathname for load map: Input/output error.

warning: core file may not match specified executable file.
[New LWP 3319]

warning: Can't read pathname for load map: Input/output error.
Core was generated by `./forever4'.
Program terminated with signal 11, Segmentation fault.
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3319):
#0  0x00000000004004b8 in main ()
Core was generated by `./forever4'.
Program terminated with signal 11, Segmentation fault.
#0  0x00000000004004b8 in main ()

Thread 1 (LWP 3319):
#0  0x00000000004004b8 in main ()

[15]   Done                    stop_it forever
[16]   Done                    stop_it forever2
[18]   Done                    stop_it forever4
```

**Comment by springmeyer** [comment]
Proposed solution at #6

**Comment by springmeyer** [comment]
Solved by #6
